### PR TITLE
Implement Matrix class.

### PR DESCRIPTION
**Pull request by saileshsimhadri**

### CODE REVIEW

**Comment by josephsweeney** [comment]
Can we rename `determinant()` to `det()`? that's gonna get real old real fast.

**Comment by josephsweeney** [comment]
We also need scalar multiplication.

**Comment by josephsweeney** [comment]
which means you'll need an `__lmul__` as well.

**Comment by DrewMonroe** [comment]
Did we want to have any kind of matrix power operations? So like `A^2`?

**Comment by saileshsimhadri** [comment]
I could add it in. Do you guys want me to?

File: matrix.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        """Return numpy array's str."""
+        return str(self._columns)
+
+    def __str__(self):
+        """Return numpy array's str."""
+        return str(self._columns)
+
+    def width(self):
+        """Returns the width of the matrix"""
+        return len(self._colums[0])
```
Comment (DrewMonroe) [inline matrix.py]: this should read `return len(self._columns[0])`

**Comment by DrewMonroe** [comment]
I can go either way with it, it might be useful, but we could always implement it if we need it

**Comment by josephsweeney** [comment]
Let's hold off and add it later. once the matrix is in we can add some predicates which we need.

File: matrix.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        return self._columns
+
+    def __mul__(self, other):
+        """Multiply this matrix with another matrix"""
+        if isinstance(other, (int, float)):
+            return Matrix(*[Vector(*i) for i in
+                            np.multiply(self._columns, other)])
+        elif type(self) == type(other):
+            return Matrix(*[Vector(*i) for i in
+                        self._columns.dot(other.to_array()).T])
```
Comment (josephsweeney) [inline matrix.py]: this is under-indented

Comment (josephsweeney) [reply matrix.py]: It needs two more spaces.

**Review by josephsweeney — CHANGES_REQUESTED** [review]

File: matrix.py (hunk 2 above)
Comment (josephsweeney) [reply matrix.py]: specifically line 47 cause its kind of vague what i'm referring to.

**Comment by saileshsimhadri** [comment]
Didn't see your post. I implemented it.

**Comment by josephsweeney** [comment]
Line #75 is too long

**Comment by josephsweeney** [comment]
Thats fine.

File: matrix.py (hunk 2 above)
Comment (saileshsimhadri) [reply matrix.py]: what is under indented?

Comment (saileshsimhadri) [reply matrix.py]: never mind got it

File: matrix.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        """Checks if two matrices are equal"""
+        return np.array_equal(self._columns, other.to_array())
+
+    def transpose(self):
+        """Returns the transpose of the matrix"""
+        return Matrix(*[Vector(*i) for i in self._columns])
+
+    def power(self, p):
+        """Returns a matrix to the power p"""
+        return Matrix(*[Vector(*i) for i in np.linalg.matrix_power(self._columns, p).T])
```
Comment (DrewMonroe) [inline matrix.py]: As @josephsweeney it's too long of a line

**Review by DrewMonroe — CHANGES_REQUESTED** [review]

**Review by josephsweeney — APPROVED** [review]

File: matrix.py (hunk 3 above)
Comment (saileshsimhadri) [reply matrix.py]: Changed it

**Review by DrewMonroe — APPROVED** [review]

**Pull request merged**